### PR TITLE
SN1-138 Refactoring Prompting API

### PR DESCRIPTION
**Pull request by Hollyqui**
# Changes

This refactor includes the following changes:

- `Poetry` - We now use poetry for installing dependencies
- `.env file` - We no longer pass in arguments and instead use a .env file (this decouples the bittensor config as well)
- `Schemas` - The API schemas (request params as well as response) have changed to match the changes necessary 
- `Prompting repo` - the API is now standalone and there is no more dependency on the prompting repo.  Other things like the dashboard and API features from SN25 were removed.
- `Dashboard` - The dashboard was removed from this repo and moved to its own repo: https://github.com/macrocosm-os/prompting-dashboard
- `Streaming` - The code now streams chunk-by-chunk instead of processing all the data and then streaming.  This makes for a better user experience.
  - Because of this change, we can no longer only return the longest text response since that requires receiving all responses, processing, and then streaming.
- `Validator or Miner` - The API now supports querying miners or validators based on the request’s parameters and will stream k number of miner’s responses through the API.
- `Error Handling` - Better error handling was added.  This includes handling of timeouts and incapable request inputs (e.g. specifying a sample mode of “list” and not providing a list of UIDs to sample from.
- `Metagraph resync` - The metagraph is automatically resync’d every 60 seconds (or as defined by the .env) via an asynchronous task

# Detailed Test Plan
https://docs.google.com/document/d/1TJvl8zsl3_Badekq0jhBv4pxVjWvTAQQ7kdsuANYMkE/edit

# Future PRs

- Add unit tests
- Create a master API that will work across subnets that the prompting-api can inherit from

### CODE REVIEW

File: network/meta/protocol.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+# and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
```
Comment (steffencruz) [inline network/meta/protocol.py:16]:
```suggestion
```

File: network/meta/protocol.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+import pydantic
+import bittensor as bt
+from typing import List, AsyncIterator
+from starlette.responses import StreamingResponse
+
+
+class StreamPromptingSynapse(bt.StreamingSynapse):
```
Comment (steffencruz) [inline network/meta/protocol.py:24]: Is this a full duplicate of the synapse that is defined in the prompting repo?

Comment (cassova) [reply network/meta/protocol.py:8]: yes, this is pretty much copy paste.

**Review by bkb2135 — APPROVED** [review]
The code looks good and the tests seem very sensible.

File: network/meta/middlewares.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+#     try:
+#         # Parsing JSON data from the request
+#         request["data"] = await request.json()
+#     except json.JSONDecodeError as e:
+#         bt.logging.error(f"Invalid JSON data: {str(e)}")
+#         return Response(status=400, text="Invalid JSON")
+
+#     # Continue to the next handler if JSON is successfully parsed
+#     return await handler(request)
```
Comment (dbobrenko) [inline network/meta/middlewares.py:51]: Do we need this commented code?

File: network/stream_manager.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        self.accumulated_timings = defaultdict(list)
+        self.start_time = time.perf_counter()
+
+        try:
+            async with async_timeout.timeout(self.request.timeout):
+                for stream_response, uid in zip(streams_responses, stream_uids):
+                    # Determine which UID was passed in
+                    validator_uid = uid if self.request.query_validators else -1
+                    miner_uid = uid if not self.request.query_validators else -1
+                    logger.info(f"Miner UID: {miner_uid}, Validator UID: {validator_uid}")
```
Comment (dbobrenko) [inline network/stream_manager.py:46]: Should we inform user more which mode is used? Logging UID: -1 can be confusing

**Review by dbobrenko — APPROVED** [review]
LGTM, nicely done!

File: network/meta/middlewares.py (hunk 3 above)
Comment (cassova) [reply network/meta/middlewares.py:51]: good point. I'll take that out...

**Pull request merged**